### ISSUE TITLE
Support for dynamically changing language

### ISSUE THREAD

**Issue by Procyon-b** [cxw42/TabFern#206]
I have added the possibility for the user to select the interface language of his choice.

You can download my "select-lang" branch. It is based on "rc" which apparently is the current released code.
3 files have been modified to add options, and to handle the use of the correct "message.json" files.

**Settings**:
- the language option is added under the "appearance" section. The current supported languages are listed in the drop-down select. (the charset of file `manifest.js` is now UTF-8 (no BOM) instead of ANSI to support the value "Fran`ç`ais" )
- "Browser default" mimicks how i18n find the matching language based on your chrome settings (interface language + accepted languages).
- "[ User defined ]" is something I've already tried in the beta version of one of my extensions. You paste the content of a "test" message.json into the textarea element, and it tries to use that.
As a test paste this, reload the "Settings" page, and the H1 title will change ("[ User defined ]" must be selected).
```
{
"wsSettings": {
    "message": "TEST !!!"
    ,"description":"The title for the Settings window"
  }
}
```
The textarea is ugly, But I couldn't style it as I wished (not tried much). In my extension, I display an "Edit" button only when "[ User defined ]" is selected. When clicked this displays the textarea element.
I can make the beta version of that extension available for download if you want to see a usable, user-friendly, interface. In it, the setting are applied immediately. All pages of the extension are reloaded, displaying the correct language immediately.

**Logic**
From what I've read recently (I can try to dig up the source if you want), `i18n.getMessage()` works like this:
- tries to load the message from your locale
- if no message, tries to load from the "basic" locale of you locale (en-US -> en)
- if no message, tries to load from the default locale (as per manifest.json setting)

So, what my code does, is first to find an existing locale which matches the user settings in this order:
`[lang || '', chrome.i18n.getUILanguage()].concat(navigator.languages).concat([defaultLang])`
and in the process, try the "basic" locale for each "en-US" syntax.

The first match will be the default locale.
It completes the locale with elements from the "basic" locale (if the match is eg. "en-US"), and with the default locale (as per manifest.json definition).

`_T` is linked to a function that gets the message from the "generated" locale, replacing "placeholders" if necessary.
Note that it handles "placeholders" correctly for the syntax used in TF, but not yet for the all the possibilities (only found in one of the extensions installed in my profiles). It's not difficult at all to implement.

**Comment by cxw42** [comment]
Sounds interesting!  Would you please open a PR so we can discuss there?  Thanks!

**Comment by cxw42** [comment]
Haven't had a chance yet, but I plan to

**Comment by cxw42** [comment]
Related to https://github.com/cxw42/TabFern/issues/135#issuecomment-415035168

**Comment by Procyon-b** [comment]
By the way, I forgot to mention that I'm unable to access `https://github.com/frankkohlhepp/fancy-settings`.
Is it a private repository?

**Comment by cxw42** [comment]
No, as far as I know fancy-settings has disappeared.  I got it from a copy of the code in the repo I used as a template for this extension ([here](https://github.com/altryne/extensionizr/blob/master/ext/src/options_custom/README.md)).

**Comment by cxw42** [comment]
I looked at your changes and generally do not object.  I would like to discuss in a PR whenever you have a chance to open one.  Thanks!

**Comment by Procyon-b** [comment]
I'll submit a PR. I'll also look at the source of fancy-settings to see what king of syntax it accepts.

I find the possibility to change the language more interesting now that in my extension (beta) I've added the possibility for the user to add his own locale. It can help the translator to test the conversion of the locale to his own language. And it can be useful for new users when a locale exists  for a language but is not yet included in the published version of the extension.


**CONTRIBUTING.md**
Parts of your explanation are missing newlines. From experience, `<br>` works.

**Comment by Procyon-b** [comment]
I've updated my branch.
It is base on your "rc" which has now disappeared (merged in master ?). Do I create the PR on "master"?

Here is a list of modifications so far (I'll repeat in the PR):
- updated `manifest.js` to support localization: `future_i18n()` tries to get localized string from `_T()`
- A bunch of strings in `manifest.js` now call `future_i18n()` with 1 or more optional parameters: second arg is the name of the message, and other args are placeholders (used a couple times). `future_i18n()` is still compatible with old usage to support messages not yet localized.
- Newly localized "settings" page's messages are not included in the EN locale (since they are defined in `setting_definitions`).
- FR locale contains all text from the "settings" page.
- settings page: the "language" block under "appearance" has been refined. The textarea for the user-defined locale is hidden by default. A button appears next to the "select" element only when "user defined" is selected. Clicking this button toggles the display of the textarea.
- To handle this behaviour, `__custom.css` contains style to move the button next to the select (float) and to hide these elements.
code in `setting.js` adds CSS classes to these elements (CSS classes can't be added by editing `setting_definitions`), and defines event-based functions to handle user actions (using CSS classes).

**Comment by cxw42** [comment]
Yes, I merged `rc` into `master`.  No worries --- if there are any conflicts, we can fix them in the PR.  Sounds like a good chunk of work --- I really appreciate it!

**Comment by Procyon-b** [comment]
I've encountered a problem with `npx brunch w` it didn't caught the modifications of `settings/__custom.css`.
I had to stop/start the process each time. It took me a minute or two to realize.  ;)
